File: internal/engine/runtime.go
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package engine

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"slices"
	"time"

	"github.com/NVIDIA/NemoClaw/internal/config"
	"github.com/NVIDIA/NemoClaw/internal/managed"
	oshell "github.com/NVIDIA/NemoClaw/internal/openshell"
)

func runtimeSpecs(d config.Document, g map[string]string) []managed.Spec {
	if d.Spec.Gateway.Management != "managed" {
		return nil
	}
	s := []managed.Spec{{Layout: 2, Kind: managed.GatewayKind, Name: d.Workspace() + "-gateway", Owner: d.Metadata.UID, Generation: g[managed.GatewayKind], Gateway: d.Spec.Gateway}}
	if service := d.Spec.InferenceProviders[0].Service; service != nil {
		s = append(s, managed.Spec{Kind: managed.ServiceKind, Name: d.Workspace() + "-inference", Owner: d.Metadata.UID, Generation: g[managed.ServiceKind], Gateway: d.Spec.Gateway, Service: service})
	}
	return s
}

func runtimeAddress(s managed.Spec) string { return "nemoclaw_" + s.Kind + ".runtime" }

func runtimeStorage(d config.Document, g map[string]string) managed.Storage {
	return managed.Storage{Name: d.Workspace() + "-inference-data", Owner: d.Metadata.UID, Generation: g[managed.ServiceKind], Engine: d.Spec.Gateway.Engine}
}

func gatewayStorageSpec(d config.Document, g map[string]string) managed.Spec {
	s := runtimeSpecs(d, g)[0]
	s.Layout = 0
	return s
}

const gatewayStorageAddress = "nemoclaw_gateway_storage.runtime"
const storageAddress = "nemoclaw_inference_storage.runtime"

func (e *Engine) runtimeStage(ctx context.Context, operation string, d config.Document, r *Record) ([]Change, bool, error) {
	if d.Spec.Gateway.Management != "managed" {
		return nil, false, nil
	}
	for _, k := range []string{managed.GatewayKind, managed.ServiceKind} {
		if r.Generations[k] == "" {
			r.Generations[k] = newGenerations()[k]
		}
	}
	stage := &Engine{StateDir: filepath.Join(e.StateDir, "runtime"), BundleDir: e.BundleDir, Output: e.Output}
	if err := os.MkdirAll(stage.StateDir, 0700); err != nil {
		return nil, false, err
	}
	ids, err := stage.stateIDs()
	if err != nil {
		return nil, false, err
	}
	bindings, err := stage.stateBindings()
	if err != nil {
		return nil, false, err
	}
	specs := runtimeSpecs(d, r.Generations)
	expected := len(specs) + 1
	if d.Spec.InferenceProviders[0].Service != nil {
		expected++
	}
	if len(ids) > expected {
		return nil, false, errors.New("ordinary apply cannot remove a managed runtime")
	}
	docker, err := managed.New(d.Spec.Gateway.Engine)
	if err != nil {
		return nil, false, err
	}
	defer docker.Close()
	gatewayStorageID, err := docker.GatewayStorage(ctx, gatewayStorageSpec(d, r.Generations), ids[gatewayStorageAddress], false)
	if err != nil && !errors.Is(err, managed.ErrPartial) {
		return nil, false, err
	}
	storageIdentity := ""
	if d.Spec.InferenceProviders[0].Service != nil {
		storageIdentity, err = docker.Storage(ctx, runtimeStorage(d, r.Generations), ids[storageAddress], false)
		if err != nil {
			return nil, false, err
		}
	}
	gatewayRunning := false
	replacements := map[string]bool{}
	for _, s := range specs {
		observedSpec, err := boundRuntimeSpec(s, bindings[runtimeAddress(s)])
		if err != nil {
			return nil, false, err
		}
		if observedSpec.JSON() != s.JSON() {
			switch s.Kind {
			case managed.ServiceKind:
				replacements[runtimeAddress(s)] = storageIdentity != ""
			case managed.GatewayKind:
				replacements[runtimeAddress(s)] = gatewayStorageID != ""
			}
		}
		o, err := docker.Observe(ctx, observedSpec, ids[runtimeAddress(s)])
		if err != nil && !errors.Is(err, managed.ErrPartial) {
			return nil, false, err
		}
		if s.Kind == managed.GatewayKind {
			gatewayRunning = o != nil && o.Running
		}
		if s.Service != nil {
			if err = docker.Capacity(ctx, s, o); err != nil {
				return nil, false, err
			}
		}
	}
	if err = stage.prepare(d, r.Generations); err != nil {
		return nil, false, err
	}
	b, err := os.ReadFile(filepath.Join(stage.StateDir, "main.tf.json"))
	if err != nil {
		return nil, false, err
	}
	var compiled map[string]any
	if err = json.Unmarshal(b, &compiled); err != nil {
		return nil, false, err
	}
	resources := map[string]any{"nemoclaw_gateway_storage": map[string]any{"runtime": map[string]any{"spec": gatewayStorageSpec(d, r.Generations).JSON(), "lifecycle": map[string]any{"prevent_destroy": true}}}}
	if d.Spec.InferenceProviders[0].Service != nil {
		storage, _ := json.Marshal(runtimeStorage(d, r.Generations))
		resources["nemoclaw_inference_storage"] = map[string]any{"runtime": map[string]any{"spec": string(storage), "lifecycle": map[string]any{"prevent_destroy": true}}}
	}
	for _, s := range specs {
		attrs := map[string]any{"spec": s.JSON(), "depends_on": []string{gatewayStorageAddress}}
		if s.Service != nil {
			attrs["depends_on"] = []string{"nemoclaw_managed_gateway.runtime", storageAddress}
		}
		resources["nemoclaw_"+s.Kind] = map[string]any{"runtime": attrs}
	}
	compiled["resource"] = resources
	if err = saveJSON(filepath.Join(stage.StateDir, "main.tf.json"), compiled); err != nil {
		return nil, false, err
	}
	if _, err = stage.tofu(ctx, "init", "-upgrade", "-input=false", "-no-color"); err != nil {
		return nil, false, err
	}
	if _, err = stage.tofu(ctx, "plan", "-input=false", "-no-color", "-parallelism=1", "-out=apply.plan"); err != nil {
		return nil, false, err
	}
	b, err = stage.tofu(ctx, "show", "-json", "apply.plan")
	if err != nil {
		return nil, false, err
	}
	var plan Plan
	if json.Unmarshal(b, &plan) != nil {
		return nil, false, errors.New("invalid runtime plan")
	}
	allowed := map[string]bool{gatewayStorageAddress: true}
	if d.Spec.InferenceProviders[0].Service != nil {
		allowed[storageAddress] = true
	}
	for _, s := range specs {
		allowed[runtimeAddress(s)] = true
	}
	changes, err := checkRuntimePlan(plan, allowed, replacements)
	if err != nil {
		return nil, false, err
	}
	if operation == "plan" {
		if !gatewayRunning {
			childIDs, err := e.stateIDs()
			if err != nil {
				return nil, false, err
			}
			if len(childIDs) != 0 {
				return nil, false, errors.New("runtime restart is planned but OpenShell observations are unavailable; apply the unchanged intent to reconcile the runtime stage")
			}
		}
		if r.Digest == "" {
			r.Document = d
			r.Digest = d.Digest()
			if err = saveJSON(filepath.Join(e.StateDir, "intent.json"), r); err != nil {
				return nil, false, err
			}
		}
		return changes, !gatewayRunning, nil
	}
	// Each explicit graph has its own checked saved plan. Persist intent before
	// runtime effects; a crash can then reconcile the same operation generations.
	b, err = os.ReadFile(filepath.Join(stage.StateDir, "apply.plan"))
	if err != nil {
		return nil, false, err
	}
	h := sha256.Sum256(b)
	r.Document = d
	r.Digest = d.Digest()
	r.Pending = true
	r.Succeeded = false
	r.PlanDigest = hex.EncodeToString(h[:])
	if err = saveJSON(filepath.Join(e.StateDir, "intent.json"), r); err != nil {
		return nil, false, err
	}
	if _, err = stage.tofu(ctx, "apply", "-input=false", "-no-color", "-parallelism=1", "apply.plan"); err != nil {
		return nil, false, fmt.Errorf("runtime apply incomplete; retain data and reapply identical YAML: %w", err)
	}
	r.Pending = false
	if err = saveJSON(filepath.Join(e.StateDir, "intent.json"), r); err != nil {
		return nil, false, err
	}
	if err = waitRuntime(ctx, d, specs, docker, stage); err != nil {
		return nil, false, err
	}
	return changes, false, nil
}

// State records the configuration actually established by OpenTofu. Intent may
// already contain a new image when interruption leaves the old container bound.
func boundRuntimeSpec(want managed.Spec, binding stateBinding) (managed.Spec, error) {
	if binding.ID == "" {
		return want, nil
	}
	var old managed.Spec
	if json.Unmarshal([]byte(binding.Spec), &old) != nil || old.Validate() != nil || old.Kind != want.Kind || old.Name != want.Name || old.Owner != want.Owner || old.Generation != want.Generation || old.Gateway.Engine != want.Gateway.Engine {
		return old, errors.New("managed runtime state identity or configuration is invalid")
	}
	return old, nil
}

func checkRuntimePlan(plan Plan, allowed map[string]bool, replacements map[string]bool) ([]Change, error) {
	changes := []Change{}
	seen := map[string]bool{}
	for _, c := range plan.ResourceChanges {
		replacement := (c.Address == "nemoclaw_inference_service.runtime" || c.Address == "nemoclaw_managed_gateway.runtime") && replacements[c.Address] && slices.Equal(c.Change.Actions, []string{"delete", "create"})
		if !allowed[c.Address] || seen[c.Address] || (!replacement && (len(c.Change.Actions) != 1 || !slices.Contains([]string{"no-op", "create", "update"}, c.Change.Actions[0]))) {
			return nil, errors.New("runtime plan would remove, replace, or affect an undeclared resource")
		}
		seen[c.Address] = true
		if c.Change.Actions[0] != "no-op" {
			changes = append(changes, Change{Resource: c.Address, Actions: c.Change.Actions})
		}
	}
	if len(seen) != len(allowed) {
		return nil, errors.New("runtime plan omitted a required resource; no changes applied")
	}
	return changes, nil
}

func waitRuntime(ctx context.Context, d config.Document, specs []managed.Spec, docker *managed.Docker, stage *Engine) error {
	c, err := oshell.Connect(d.Spec.Gateway)
	if err != nil {
		return err
	}
	defer c.Close()
	ready, cancel := context.WithTimeout(ctx, 90*time.Second)
	defer cancel()
	for {
		probe, stop := context.WithTimeout(ready, 2*time.Second)
		_, err = c.Health().GetGatewayInfo(probe)
		stop()
		if err == nil {
			break
		}
		select {
		case <-ready.Done():
			return errors.New("managed gateway readiness failed; established identity and data retained")
		case <-time.After(time.Second):
		}
	}
	ids, err := stage.stateIDs()
	if err != nil {
		return err
	}
	for _, s := range specs {
		if s.Service == nil {
			continue
		}
		wait, cancel := context.WithTimeout(ctx, 9*time.Hour)
		defer cancel()
		lastPhase := ""
		for {
			o, err := docker.Observe(wait, s, ids[runtimeAddress(s)])
			if err != nil {
				return err
			}
			if o == nil || !o.Running {
				return errors.New("inference runtime stopped; identity and model data retained; inspect container logs and explicitly reapply")
			}
			status, err := docker.Status(wait, o)
			if err != nil {
				return err
			}
			if status.Phase != lastPhase {
				fmt.Fprintln(os.Stderr, "Spark runtime:", status.Phase)
				lastPhase = status.Phase
			}
			if status.Phase == "ready" {
				return docker.VerifyArtifacts(wait, o)
			}
			select {
			case <-wait.Done():
				return errors.New("runtime readiness interrupted; container, watchdog, and persistent data remain owned")
			case <-time.After(5 * time.Second):
			}
		}
	}
	return nil
}

File: crates/nemoclaw-sdk/src/deployment/runtime.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
mod teardown;
#[cfg(all(test, unix))]
mod tests;

use super::*;
use crate::managed::{GATEWAY_KIND, GATEWAY_STORAGE_KIND, Spec};
const GATEWAY_STORAGE: &str = "nemoclaw_gateway_storage.runtime";
pub(super) fn check_runtime_plan(
    plan: &Plan,
    allowed: &BTreeMap<String, Row>,
    bindings: &BTreeMap<String, StateBinding>,
    replacements: &BTreeSet<String>,
) -> Result<Vec<Change>, Error> {
    let mut ordinary = Plan::default();
    let mut changes = Vec::new();
    let mut seen = BTreeSet::new();
    let mut observations = BTreeSet::new();
    for change in &plan.resource_changes {
        if plan::observation(change, &mut observations, allowed, true, false)? {
            continue;
        }
        if plan::disposable(&change.address) || change.deposed.is_some() {
            ordinary.resource_changes.push(plan::ResourceChange {
                mode: change.mode.clone(),
                address: change.address.clone(),
                deposed: change.deposed.clone(),
                change: plan::PlannedChange {
                    actions: change.change.actions.clone(),
                    before: change.change.before.clone(),
                    after: change.change.after.clone(),
                },
            });
            continue;
        }
        if !seen.insert(&change.address) {
            return Err(Error::Conflict("runtime plan duplicated a resource"));
        }
        let expected = allowed.get(&change.address).ok_or(Error::Conflict(
            "runtime plan contains an undeclared resource",
        ))?;
        if change.change.actions == ["delete", "create"] {
            if !replacements.contains(&change.address) {
                return Err(Error::Conflict(
                    "runtime replacement requires verified retained storage",
                ));
            }
            let binding = bindings
                .get(&change.address)
                .ok_or(Error::Conflict("runtime replacement is unbound"))?;
            if change.change.before["id"] != binding.id
                || change.change.before["spec"] != expected["spec"]
            {
                return Err(Error::Conflict(
                    "runtime replacement changed the established identity or specification",
                ));
            }
            changes.push(Change {
                resource: change.address.clone(),
                actions: change.change.actions.clone(),
            });
            ordinary.resource_changes.push(plan::ResourceChange {
                mode: change.mode.clone(),
                address: change.address.clone(),
                deposed: change.deposed.clone(),
                change: plan::PlannedChange {
                    actions: vec!["no-op".into()],
                    before: change.change.before.clone(),
                    after: change.change.after.clone(),
                },
            });
        } else {
            ordinary.resource_changes.push(plan::ResourceChange {
                mode: change.mode.clone(),
                address: change.address.clone(),
                deposed: change.deposed.clone(),
                change: plan::PlannedChange {
                    actions: change.change.actions.clone(),
                    before: change.change.before.clone(),
                    after: change.change.after.clone(),
                },
            });
        }
    }
    changes.extend(check_plan(&ordinary, allowed, bindings)?);
    Ok(changes)
}
fn bound_spec(want: &Spec, binding: Option<&StateBinding>) -> Result<Spec, Error> {
    let Some(binding) = binding else {
        return Ok(want.clone());
    };
    let old: Spec = serde_json::from_str(&binding.spec)
        .map_err(|_| Error::Conflict("bound runtime specification is incomplete"))?;
    old.validate()?;
    if old.kind != want.kind
        || old.name != want.name
        || old.owner != want.owner
        || old.generation != want.generation
        || old.engine() != want.engine()
    {
        return Err(Error::Conflict(
            "bound runtime identity differs from retained intent",
        ));
    }
    Ok(old)
}
struct RuntimeValidation {
    expected: BTreeMap<String, Row>,
    replacements: BTreeSet<String>,
    gateway_running: bool,
}
// Binding validation is local. Live identity and running state come from the
// provider refresh in the saved plan, never from a separate SDK preflight.
fn runtime_bindings(
    targets: &[Target],
    bindings: &BTreeMap<String, StateBinding>,
) -> Result<BTreeMap<String, Row>, Error> {
    let mut expected = allowed(targets);
    if bindings.keys().any(|key| {
        !expected.contains_key(key) && (!plan::disposable(key) || key.starts_with("docker_volume."))
    }) {
        return Err(Error::Conflict(
            "ordinary apply cannot remove a managed runtime",
        ));
    }
    for target in targets {
        if target.kind == GATEWAY_KIND
            && plan::disposable(&target.address)
            && bindings.contains_key(&target.address)
            && !bindings.contains_key(GATEWAY_STORAGE)
        {
            return Err(Error::Conflict(
                "gateway compute requires its retained storage binding",
            ));
        }
        if plan::disposable(&target.address) || target.address.starts_with("data.") {
            continue;
        }
        if target.kind == GATEWAY_KIND
            || crate::services::resource_behavior(&target.kind).runtime_process
        {
            let want: Spec = serde_json::from_str(&target.values["spec"])
                .map_err(|_| Error::State("invalid compiled runtime"))?;
            let old = bound_spec(&want, bindings.get(&target.address))?;
            expected
                .get_mut(&target.address)
                .unwrap()
                .insert("spec".into(), old.json()?);
        } else if bindings
            .get(&target.address)
            .is_some_and(|binding| binding.spec != target.values["spec"])
        {
            return Err(Error::Conflict(
                "bound storage specification differs from retained intent",
            ));
        }
    }
    Ok(expected)
}
fn runtime_observations(
    document: &Document,
    generations: &crate::compile::Generations,
    targets: &[Target],
    bindings: &BTreeMap<String, StateBinding>,
    plan: &Plan,
) -> Result<RuntimeValidation, Error> {
    let mut result = RuntimeValidation {
        expected: runtime_bindings(targets, bindings)?,
        replacements: BTreeSet::new(),
        gateway_running: document.spec.gateway.as_managed().is_none(),
    };
    let retained: BTreeSet<_> = targets
        .iter()
        .filter(|target| {
            target.kind == GATEWAY_STORAGE_KIND
                || crate::services::resource_behavior(&target.kind).retained_storage
        })
        .filter(|target| {
            bindings.get(&target.address).is_some_and(|binding| {
                plan.resource_changes.iter().any(|change| {
                    change.address == target.address
                        && change.mode.as_deref() != Some("data")
                        && change.change.actions == ["no-op"]
                        && change.change.before["id"] == binding.id
                        && change.change.before["spec"] == binding.spec
                })
            })
        })
        .map(|target| target.address.clone())
        .collect();
    if let Some(gateway) = targets
        .iter()
        .find(|target| target.kind == GATEWAY_KIND && plan::disposable(&target.address))
    {
        result.gateway_running = plan.resource_changes.iter().any(|change| {
            change.address == gateway.address
                && change.change.actions == ["no-op"]
                && change.change.before["id"]
                    .as_str()
                    .is_some_and(|id| !id.is_empty())
        });
    }
    for target in targets.iter().filter(|target| {
        !plan::disposable(&target.address)
            && (target.kind == GATEWAY_KIND
                || crate::services::resource_behavior(&target.kind).runtime_process)
    }) {
        if target.kind == GATEWAY_KIND {
            result.gateway_running = plan.resource_changes.iter().any(|change| {
                change.address == target.address && change.change.before["running"] == "true"
            });
        }
        let storage = if target.kind == GATEWAY_KIND {
            Some(GATEWAY_STORAGE.to_owned())
        } else {
            crate::services::required_storage_address(document, generations, &target.address)?
        };
        if result.expected[&target.address]["spec"] != target.values["spec"]
            && storage.is_some_and(|address| retained.contains(&address))
        {
            result.replacements.insert(target.address.clone());
        }
    }
    Ok(result)
}

impl Deployment {
    pub(super) async fn runtime_stage(
        &self,
        bundle: &Bundle,
        store: &Store,
        document: &Document,
        record: &mut Record,
        apply: bool,
        cancel: &CancellationToken,
    ) -> Result<(Vec<Change>, bool), Error> {
        if !document.has_runtime() {
            let directory = store.directory.join("runtime");
            if directory.exists()
                && !self
                    .state_bindings(
                        bundle,
                        &Store::open(&directory)?,
                        &record.document,
                        &record.generations,
                        true,
                        cancel,
                    )
                    .await?
                    .is_empty()
            {
                return Err(Error::Conflict(
                    "a configuration without runtime requires a new state directory; retain the existing runtime configuration and state for recovery or destroy",
                ));
            }
            return Ok((Vec::new(), false));
        }
        let generated = Record::new(document.clone())?;
        for (kind, generation) in generated.generations {
            if record.generations.get(&kind).is_none_or(String::is_empty) {
                record.generations.insert(kind, generation);
            }
        }
        let stage = Store::open(&store.directory.join("runtime"))?;
        let bindings = self
            .state_bindings(bundle, &stage, document, &record.generations, true, cancel)
            .await?;
        let targets = compile::runtime_targets(document, &record.generations)?;
        runtime_bindings(&targets, &bindings)?;
        self.prepare(
            bundle,
            &stage,
            &compile::compile_runtime(document, &record.generations, &bundle.manifest.version)?,
        )?;
        self.tofu(
            bundle,
            &stage,
            document,
            &["init", "-upgrade", "-input=false", "-no-color"],
            cancel,
        )
        .await?;
        let plan = self
            .saved_plan(bundle, &stage, document, "apply.plan", cancel)
            .await?;
        let checked =
            runtime_observations(document, &record.generations, &targets, &bindings, &plan)?;
        let changes =
            check_runtime_plan(&plan, &checked.expected, &bindings, &checked.replacements)?;
        if !apply {
            if !checked.gateway_running
                && !self
                    .state_bindings(bundle, store, document, &record.generations, false, cancel)
                    .await?
                    .is_empty()
            {
                return Err(Error::Conflict(
                    "the managed gateway is not running, so plan cannot inspect OpenShell resources; run apply with the same configuration and state directory to restore the gateway",
                ));
            }
            store.save(record)?;
            return Ok((changes, !checked.gateway_running));
        }
        record.document = document.clone();
        record.digest = document.digest();
        record.begin_runtime_apply();
        record.succeeded = false;
        record.destroyed = false;
        record.destroy_runtime = false;
        record.plan_digest = crate::bundle::hash_file(&stage.directory.join("apply.plan"))?;
        store.save(record)?;
        self.tofu(
            bundle,
            &stage,
            document,
            &["apply", "-input=false", "-no-color", "apply.plan"],
            cancel,
        )
        .await?;
        record.finish_runtime_apply();
        store.save(record)?;
        Ok((changes, false))
    }
    pub(super) async fn export_runtime(
        &self,
        bundle: &Bundle,
        store: &Store,
        record: &Record,
        cancel: &CancellationToken,
    ) -> Result<(), Error> {
        if !record.document.has_runtime() {
            return Ok(());
        }
        let stage = Store::open(&store.directory.join("runtime"))?;
        self.export_observations(bundle, &stage, record, true, cancel)
            .await?;
        Ok(())
    }
}
